### PR TITLE
feat(ApplicationCommand): add support for channel_types

### PR DESCRIPTION
**Pull request by ckohen**
**Please describe the changes this PR makes and why it should be merged:**

🚀 

- https://github.com/discord/discord-api-docs/pull/3825

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

File: src/structures/ApplicationCommand.js [hunk 1]
```diff
@@ -131,6 +131,12 @@ class ApplicationCommand extends Base {
    * @property {boolean} [required] Whether the option is required
    * @property {ApplicationCommandOptionChoice[]} [choices] The choices of the option for the user to pick from
    * @property {ApplicationCommandOptionData[]} [options] Additional options if this option is a subcommand (group)
+   * @property {ChannelType[]|number[]} [channelTypes] When the option type is channel,
+   * the allowed types of channels that can be selected
+   * @property {number[]} [channel_types] When the option type is channel,
+   * the API data for allowed types of channels that can be selected
+   * <warn>This is provided for compatability with something like `@discordjs/builders`
```
Comment (ThaTiemsz) [inline src/structures/ApplicationCommand.js:138]:
```suggestion
   * <warn>This is provided for compatibility with something like `@discordjs/builders`
```

**Review by vladfrangu — APPROVED** [review]
The warn block could use some rephrasing, but otherwise LGTM

File: typings/index.d.ts [hunk 2]
```diff
@@ -3077,10 +3083,11 @@ export interface ApplicationCommandNonOptionsData extends BaseApplicationCommand
 export type ApplicationCommandOptionData =
   | ApplicationCommandSubGroupData
   | ApplicationCommandNonOptionsData
+  | ApplicationCommandChannelOptionData
   | ApplicationCommandChoicesData
   | ApplicationCommandSubCommandData;
 
-export type ApplicationCommandOption = ApplicationCommandOptionData & {
+export type ApplicationCommandOption = Omit<ApplicationCommandOptionData, 'ApplicationCommandChannelOptionRawData'> & {
```
Comment (almeidx) [inline typings/index.d.ts:3090]: What is `ApplicationCommandChannelOptionRawData` supposed to be?

Comment (ckohen) [reply typings/index.d.ts:3090]: Oh, that's residual, do need to fix what that was trying to though (the ApplicationCommandOption type shouldn't have channel_types ever)

Comment (almeidx) [reply typings/index.d.ts:3090]: `Omit` is used to remove certain keys from a type. You were likely looking for `Exclude`, which is used for removing a type from a union.

Although, you could just omit the `channel_types` key🤷

Comment (ckohen) [reply typings/index.d.ts:3090]: When it existed it was Exclude, forget to change the key and just put quotes around it 💀. But omitting doesn't work as intended, will figure something out.

Comment (ckohen) [reply typings/index.d.ts:3090]: On second thought, just gonna rewrite this whole union. It causes options to show up on all types when it shouldn't anyways

File: src/structures/ApplicationCommand.js [hunk 3]
```diff
@@ -239,7 +245,8 @@ class ApplicationCommand extends Base {
       (option.required ?? (['SUB_COMMAND', 'SUB_COMMAND_GROUP'].includes(optionType) ? undefined : false)) !==
         existing.required ||
       option.choices?.length !== existing.choices?.length ||
-      option.options?.length !== existing.options?.length
+      option.options?.length !== existing.options?.length ||
+      (option.channelTypes?.length ?? option.channel_types.length) !== existing.channelTypes?.length
```
Comment (NotSugden) [inline src/structures/ApplicationCommand.js:249]:
```suggestion
      ((option.channelTypes ?? option.channel_types).length) !== existing.channelTypes?.length
```

Comment (ckohen) [reply src/structures/ApplicationCommand.js:249]: I was never really a fan of this pattern, but in this case it works.

**Review by vladfrangu — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Pull request merged**